---
title: "Dataviz challenge - Dan Krochmal"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part A

```{r load libraries and files}
library(ggplot2)
library(cowplot)
library(tidyverse)
source("readpheno.R")
```

```{r part A.1}

#Generate histogram
plot1 <- ggplot(pheno, aes_string(x = "TA")) +
  geom_histogram(binwidth = 5) + 
  theme_cowplot()
plot1

#Count outliers
TA.cutoffs <- quantile(pheno$TA, na.rm = T, probs = c(.25, .75)) + c(-1.5, 1.5) * IQR(pheno$TA, na.rm = T)
filter(pheno, TA < TA.cutoffs[1] | TA > TA.cutoffs[2]) %>%
  pull(TA) %>%
  sort %>%
  length
```

**The distribution is roughly normal. There are some unusually small values. Based on IQR rule there are 28 outliers**

```{r Part A.2}
#Create a pheno subset with removed outliers
pheno_ro <- subset(pheno, TA > TA.cutoffs[1] & TA < TA.cutoffs[2])

#Calculate linear regression coefficients for pheno with and without outliers
lm1 <- lm(tibia ~ TA, pheno)
lm2 <- lm(tibia ~ TA, pheno_ro)

#Scatter plot for pheno with outliers
plot2 <- ggplot(pheno, aes_string(x = "TA", y = "tibia")) +
  geom_point() +
  geom_abline(color = "blue", intercept = coef(lm1)[1], slope = coef(lm1)[2]) +
  theme_cowplot()

plot2

#Scatterplot for pheno without outliers
plot3 <- ggplot(pheno_ro, aes_string(x = "TA", y = "tibia")) +
  geom_point() +
  geom_abline(color = "blue", intercept = coef(lm2)[1], slope = coef(lm2)[2]) +
  theme_cowplot()

plot3

print(lm1)
print(lm2)
```
**The mass of the TA muscle seems to be increasing with the tibia length. Outliers don't seem to affect the output coefficients of linear regression**

```{r Part A.3}

#Generate histogram of AvToneD3 values
plot4 <- ggplot(pheno, aes_string(x = "AvToneD3")) +
  geom_histogram() + #The distribution is not normal
  theme_cowplot()

plot4

#Creating a backup of the pheno dataset
pheno2 <- pheno

#Defining logit
logit <- function(x) log((x + 0.001) / (1 - x + 0.001))

#Convert values with logit
pheno2$AvToneD3 <- logit(pheno2$AvToneD3)

#Re-do the plot4 histogram with converted values
plot5 <- ggplot(pheno2, aes_string(x = "AvToneD3")) +
  geom_histogram() +
  theme_cowplot()

plot5

#Generate boxplot
plot6 <- ggplot(pheno2, aes_string(x = "FCbox", y = "AvToneD3")) +
  geom_boxplot() +
  theme_cowplot()

plot6
```

**The distribution of AvToneD3 values isn't normal before using logit function. However transformation with logit yields distribution of values closer to normal. The type of apparatus used to perform the tone measurements seems to affect the results to some degree**


## Part B

```{r Part B}
plot7 <- ggplot(pheno, aes_string(x = "BMD")) +
  geom_histogram() +
  theme_cowplot()

plot7

hmdp <- read.csv("hmdp.csv", stringsAsFactors = FALSE)
hmdp$femur_convert <- hmdp$femur * 1000

plot8 <- ggplot() +
  geom_histogram(data = pheno, mapping = aes_string(x = "BMD"), binwidth = 5, fill = "slateblue", alpha = .4) +
  geom_histogram(data = hmdp, mapping = aes_string(x = "femur_convert"), bindwidth = 5, fill = "red3", alpha = .4) +
  #legend("topright", c("Germany", "Plastic"), col = c("slateblue", "red3"), lwd=10) +
  labs(
    title = "BMD distribution in CFW vs HMDP mice",
    x = "Bone-mineral density (mg/cm2)",
    y = "Count"
  ) +
  theme_cowplot()

plot8
```

## Part C

```{r Part C.1 prepare dataframe}
gwscan <- read.csv("gwscan.csv", stringsAsFactors = FALSE)

#Compute chr size
gwscan_2 <- gwscan %>%
  group_by(chr) %>%
  summarise(chr_len = max(pos)) %>%
  
  # start pos for each chr in the cumulative coordinates
  mutate(start_cum = cumsum(as.numeric(chr_len))-as.numeric(chr_len)) %>%
  select(-chr_len) %>%
  left_join(gwscan, ., by = 'chr') %>% # add this info to original df
            
  # position of SNP in the cumkulative coordinates
  arrange(chr,pos) %>%
  mutate(pos_cum = pos + start_cum)

axis <- gwscan_2 %>%
  group_by(chr) %>%
  summarize(center = (max(pos_cum) + min(pos_cum) ) / 2 )

```

```{r Part C.1 manhattan plot}
ggplot(gwscan_2, aes(x = pos_cum, y = abnormalBMD)) +
  geom_point(aes(color = as.factor(chr), alpha = 0.75)) +
  scale_color_manual(values = rep(c("dodgerblue4", "dodgerblue2"), 19)) +
  
  scale_x_continuous(label = axis$chr, breaks = axis$center) +
  theme_cowplot() +
  theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text.x = element_text(angle = 270, vjust = 0.5),
      axis.ticks.x = element_blank(),
      axis.line.x = element_blank()
    ) +
  labs(
    x = "Chromosome",
    y = "-log10(P)"
  )

```
```{r Part C.1 strong SNPs}
strongSNPcount <- length(which(gwscan_2$abnormalBMD > 6))
print(strongSNPcount)

gwscan_QTL_cum <- gwscan_2 %>%
  select(pos_cum, abnormalBMD, chr) %>%
  filter(chr == 11, abnormalBMD > 6)

pos_cum_min <- min(gwscan_QTL_cum$pos_cum)
pos_cum_max <- max(gwscan_QTL_cum$pos_cum)
QTL_size <- (pos_cum_max - pos_cum_min) / 10^6

gwscan_QTL <- gwscan_2 %>%
  select(pos, pos_cum, chr) %>%
  filter(chr == 11, pos_cum == pos_cum_min | pos_cum == pos_cum_max)

print(QTL_size)
print(min(gwscan_QTL$pos))
print(max(gwscan_QTL$pos))
```
```{r Part C.1 Col1a1}

#From Genome UCSC
Col1a1_pos_min <- 94936224
Col1a1_pos_max <- 94953042

gwscan_Col1a1 <- gwscan_2 %>%
  select(pos, abnormalBMD, chr) %>%
  filter(chr == 11, pos > Col1a1_pos_min & pos < Col1a1_pos_max, abnormalBMD > 6)

glimpse(gwscan_Col1a1)
```


* There are 334 SNPs associated with -log10(P) larger than 6. These SNPs are located within 2 distinct regions of the genome.
* -log10(P) of 6 corresponds to the p-value of 0.000001
* QTL with the strongest association signal is located in chromosome 11. The size of this locus is roughly 2.4 Mb and it contains ~70 genes.
* The Col1a1 gene was not included in the count performed above. Maximum -log10(P) value in the region spanning this gene was equal to 4.58.


```{r Part C.2}
gwscan_2 %>%
  select(id, abnormalBMD) %>%
  filter(id == "rs29477109")

gwscan_2

geno <- read.csv("geno_rs29477109.csv", stringsAsFactors = F)

pheno3 <- pheno
pheno3$id <- as.numeric(pheno3$id)

pheno_geno <- pheno3 %>%
  left_join(geno, by = "id")

pheno_geno$dosage <- round(pheno_geno$dosage, digits = 0)

pheno_geno_2 <- pheno_geno %>%
  mutate(
    geno = ifelse(dosage == 0, "CC",
    ifelse(dosage == 1, "CT",
    ifelse(dosage == 2, "TT", NA))))

pheno_geno_2$geno <- as.factor(pheno_geno_2$geno)

ggplot(pheno_geno_2, aes(x = geno, y = BMD)) +
  geom_boxplot() +
  theme_cowplot() +
  labs(
    x = "Genotype",
    y = "BMD"
  )
```

**While there is no visible difference between BMD of CT and TT genotypes, the CC genotype presented BMD leaning toward higher values (higher 3rd quartile).** 